### PR TITLE
Add fos_rest.exception.show_exceptions config param

### PR DESCRIPTION
**Pull request by munkie**
In order to be able manually enable/disable exception in error response.
Now kernel.debug value is used and it is not useful in some cases.
For example in functional test, thou it is run in debug mode, be able to check error response as it would be in prod environment (without exception).

### CODE REVIEW

**Comment by Ener-Getick** [comment]
You can launch your functional tests as if they were in a prod environment so I'm :-1: on this

**Comment by munkie** [comment]
@Ener-Getick By functional tests i meant something like this http://symfony.com/doc/current/book/testing.html#your-first-functional-test - testing using Symfony WebTestCase and Crawler.
Most of the time tests are run with debug off (it speeds up execution).
But while developing, debug is switched on, so cache could be rebuild on every code change (not to run cache clear every time).
In this case error response JSON assertions will fail because they will also contain exception.

**Comment by lsmith77** [comment]
also PR needs a rebase

**Comment by lsmith77** [comment]
but I like the idea

**Comment by munkie** [comment]
@lsmith77 @Ener-Getick Rebased master.
Also added _fos_rest.exception.debug_message_ option to enable/disable usage of exception message in error response manually.

**Comment by munkie** [comment]
Rebased one more time

File: DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -388,6 +403,14 @@ private function addExceptionSection(ArrayNodeDefinition $rootNode)
                             ->useAttributeAsKey('name')
                             ->prototype('boolean')->end()
                         ->end()
+                        ->booleanNode('show_exception')
+                            ->treatNullLike($this->debug)
+                            ->defaultValue($this->debug)
+                        ->end()
+                        ->booleanNode('debug_message')
```
Comment (Ener-Getick) [inline DependencyInjection/Configuration.php]: I don't see any benefits to have two fields.

File: DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -388,6 +403,14 @@ private function addExceptionSection(ArrayNodeDefinition $rootNode)
                             ->useAttributeAsKey('name')
                             ->prototype('boolean')->end()
                         ->end()
+                        ->booleanNode('show_exception')
+                            ->treatNullLike($this->debug)
```
Comment (Ener-Getick) [inline DependencyInjection/Configuration.php]: This conversion should be done in the Extension instead IMO.

File: DependencyInjection/Configuration.php (hunk 1 above)
Comment (munkie) [inline DependencyInjection/Configuration.php]: One _show_exception_ forces to add exception to error response
Second _debug_message_ forces to use exception message instead of standard response status text
Maybe second config param should renamed to _show_exception_message_ to be more clear

File: DependencyInjection/Configuration.php (hunk 2 above)
Comment (munkie) [inline DependencyInjection/Configuration.php]: Dont see much benefits in moving to extension:
```
$showException = (null === $config['exception']['show_exception']) ? $container->getParameter('kernel.debug') : $config['exception']['show_exception'];
$container->getDefinition('fos_rest.exception.controller')->replaceArgument(2, $showException);
```
Looks way more complicated.
Actually i used same approach as in symfony docs - http://symfony.com/doc/current/cookbook/configuration/using_parameters_in_dic.html

Comment (Ener-Getick) [inline DependencyInjection/Configuration.php]: ``$container->getParameter('kernel.debug')`` can be inlined in the xml file.
And the rest can be replaced by
```php
if (null !== $config['exception']['show_exception']) {
   $container->getDefinition('fos_rest.exception.controller')->replaceArgument(2, $config['exception']['show_exception']);
}
```

IMO that's clearer and that way we do not need to add a construtor argument to our ``Configuration`` class.

**Comment by Ener-Getick** [comment]
As there is not a lot of use cases, I'm wondering if it wouldn't be easier to have a container parameter aliasing ``kernel.debug``.

**Comment by munkie** [comment]
I noticed that now exception message would not be added to json/xml response with debug flag on, cause normalizer add only two fields to result data: code and message

**Comment by munkie** [comment]
@Ener-Getick @lsmith77 So whats up with this PR ?

File: DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -388,6 +388,7 @@ private function addExceptionSection(ArrayNodeDefinition $rootNode)
                             ->useAttributeAsKey('name')
                             ->prototype('boolean')->end()
                         ->end()
+                        ->booleanNode('debug')->treatNullLike(false)->end()
```
Comment (Ener-Getick) [inline DependencyInjection/Configuration.php]: IMO ``null`` should be treated as ``%kernel.debug``.

File: Tests/DependencyInjection/FOSRestExtensionTest.php [hunk 4]
```diff
@@ -603,6 +646,62 @@ public function testSerializerExceptionNormalizer()
...
+        $this->assertSame($expectedValue, $definition->getArgument(1));
+    }
+
+    /**
+     * @dataProvider getSerializerExceptionNormalizerDebugMessageData
+     *
+     * @param array|bool  $config        fost_rest.exception config
+     * @param bool|string $expectedValue Expected service debug argument value
+     */
+    public function testJmsSerializerExceptionNormalizerDebugMessage($config, $expectedValue)
```
Comment (Ener-Getick) [inline Tests/DependencyInjection/FOSRestExtensionTest.php]: Why not merging this 3 tests?

**Comment by Ener-Getick** [comment]
Except my messages fine for me :+1:

File: DependencyInjection/Configuration.php (hunk 3 above)
Comment (munkie) [inline DependencyInjection/Configuration.php]: Boolean node does not allow to use other value that boolean. Thats why i was injecting 'kernel.debug' value previously.
Without `treatNullLike(false)` _null_ will be treated as _true_

File: DependencyInjection/FOSRestExtension.php [hunk 5]
```diff
@@ -326,11 +326,20 @@ private function loadException(array $config, XmlFileLoader $loader, ContainerBu
...
+                    ->replaceArgument(1, $config['exception']['debug']);
+                $container->getDefinition('fos_rest.serializer.exception_normalizer.symfony')
+                    ->replaceArgument(1, $config['exception']['debug']);
+            }
         }
 
         foreach ($config['exception']['codes'] as $exception => $code) {
             if (!is_numeric($code)) {
-                $config['exception']['codes'][$exception] = constant("\Symfony\Component\HttpFoundation\Response::$code");
+                $config['exception']['codes'][$exception] = constant("\\Symfony\\Component\\HttpFoundation\\Response::$code");
```
Comment (xabbuh) [inline DependencyInjection/FOSRestExtension.php]: Can you revert this change please?

File: DependencyInjection/Configuration.php (hunk 3 above)
Comment (xabbuh) [inline DependencyInjection/Configuration.php]: I would also inject the `kernel.debug` parameter to the `Configuration` constructor (for example, in Symfony we do the same in the FrameworkBundle: https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L197).

File: DependencyInjection/FOSRestExtension.php [hunk 6]
```diff
@@ -326,11 +326,20 @@ private function loadException(array $config, XmlFileLoader $loader, ContainerBu
...
                 ->replaceArgument(0, $config['exception']['messages']);
+
+            if (isset($config['exception']['debug'])) {
+                $container->getDefinition('fos_rest.exception.controller')
+                    ->replaceArgument(2, $config['exception']['debug']);
+                $container->getDefinition('fos_rest.serializer.exception_normalizer.jms')
+                    ->replaceArgument(1, $config['exception']['debug']);
+                $container->getDefinition('fos_rest.serializer.exception_normalizer.symfony')
+                    ->replaceArgument(1, $config['exception']['debug']);
+            }
```
Comment (xabbuh) [inline DependencyInjection/FOSRestExtension.php]: I would not wrap this in an `if` statement. Let's just always do the replacement here.

File: Resources/config/exception_listener.xml [hunk 7]
```diff
@@ -15,7 +15,7 @@
         <service id="fos_rest.exception.controller" class="FOS\RestBundle\Controller\ExceptionController">
             <argument type="service" id="fos_rest.view_handler" />
             <argument type="service" id="fos_rest.exception.codes_map" />  <!-- exception codes -->
-            <argument>%kernel.debug%</argument>
+            <argument>%kernel.debug%</argument><!-- show exception -->
```
Comment (xabbuh) [inline Resources/config/exception_listener.xml]: Let's use something like this for clarity:

```xml
<argument /> <!-- show exception -->
```

**Comment by xabbuh** [comment]
I like this and I think we should do it in `1.8` (if I didn't miss anything this feature is fully backwards compatible).

File: DependencyInjection/Configuration.php (hunk 3 above)
Comment (Ener-Getick) [inline DependencyInjection/Configuration.php]: I told the opposite earlier but if you think this is better then why not.

Comment (xabbuh) [inline DependencyInjection/Configuration.php]: I think so as it enables us to avoid the inconsistencies we would need to discuss otherwise here.

**Comment by munkie** [comment]
@Ener-Getick @lsmith77 @xabbuh Done :sweat_smile:

File: DependencyInjection/Configuration.php [hunk 8]
```diff
@@ -29,6 +29,23 @@
...
 {
     /**
+     * kernel.debug param value.
+     *
+     * @var bool
+     */
+    private $debug;
+
+    /**
+     * Configuration constructor.
```
Comment (xabbuh) [inline DependencyInjection/Configuration.php]: I would omit this description as it doesn't add any value any just keep the parameter description.

File: DependencyInjection/Configuration.php [hunk 9]
```diff
@@ -29,6 +29,23 @@
 final class Configuration implements ConfigurationInterface
 {
     /**
+     * kernel.debug param value.
```
Comment (xabbuh) [inline DependencyInjection/Configuration.php]: This is an implementation detail which is only true when used in the extension class. We can just use something like "Whether or not to enable the debug mode".

**Comment by xabbuh** [comment]
:+1:

File: DependencyInjection/Configuration.php (hunk 9 above)
Comment (munkie) [inline DependencyInjection/Configuration.php]: Fixed :+1:

File: Resources/config/exception_listener.xml [hunk 10]
```diff
@@ -15,7 +15,7 @@
         <service id="fos_rest.exception.controller" class="FOS\RestBundle\Controller\ExceptionController">
             <argument type="service" id="fos_rest.view_handler" />
             <argument type="service" id="fos_rest.exception.codes_map" />  <!-- exception codes -->
-            <argument>%kernel.debug%</argument>
+            <argument/><!-- show exception -->
```
Comment (Ener-Getick) [inline Resources/config/exception_listener.xml]: There's missing space here (should be ``<argument />``)

Comment (munkie) [inline Resources/config/exception_listener.xml]: Fixed

**Comment by Ener-Getick** [comment]
Thank you @munkie :-)
@xabbuh can you merge this in the 1.8 branch please?

**Comment by xabbuh** [comment]
@Ener-Getick Do we want to treat this PR as a kind of bug fix?

**Comment by Ener-Getick** [comment]
@xabbuh I don't think so but 1.8 is still in beta so shouldn't we merge this PR in it ?

**Comment by Ener-Getick** [comment]
@xabbuh I created a 2.0 branch so let's just merge this in master.
